src/shillelagh/backends/apsw/db.py

```diff
@@ -500,17 +500,20 @@ def connect(
         [(1.0, 2.0), (3.0, 4.0)]
 
     """
+    if adapters is None and safe:
```

Comment:
I would say the control flow around safe is a bit hard to follow .

Reply:
Thanks for the feedback! Yeah, I agree it could be improved. It's also not perfect, as someone might be able to monkeypatch an unsafe adapter to mark it as safe. It probably need some refactoring.

BTW, did you get my email? Sometimes my email gets classified as spam because I self host it.